File: app/src/main/java/app/simone/DistributedSimon/Activities/ColorSetUpActivity.java
package app.simone.DistributedSimon.Activities;


import android.graphics.PorterDuff;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;


import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import com.google.firebase.database.ValueEventListener;

import java.util.HashMap;
import java.util.concurrent.ExecutionException;


import app.simone.R;

public class ColorSetUpActivity extends AppCompatActivity {


    private String playerID = "";
    private String playerOwnColor = "";
    private String sequenceIndex = "";
    private DatabaseReference databaseReference;
    private final String CHILD_PLAYERS = "users";
    private final String NODE_REF_ROOT = "matchesTry";
    private final String CHILD_PLAYERSSEQUENCE = "PlayersSequence";
    private final String CHILD_CPUSEQUENCE = "CPUSequence";
    private final String CHILD_MATCHID = "MATCHID";
    private final String CHILD_INDEX = "index";


    private Button buttonColor;
    private int blinkCount = 0;
    private static android.os.Handler handler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_color_set_up);
        databaseReference = FirebaseDatabase.getInstance().getReference(NODE_REF_ROOT);
        buttonColor = (Button) findViewById(R.id.beautifulButton);


        setColor();


    }

    public void sendColor(View view) throws ExecutionException, InterruptedException {
        databaseReference.child(CHILD_MATCHID).child(CHILD_PLAYERSSEQUENCE).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                long count = dataSnapshot.getChildrenCount();
                dataSnapshot.getRef().child(String.valueOf(count + 1)).setValue(playerOwnColor);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

    }

    public void onResume() {
        super.onResume();
        blink();

    }


    private void setColor() {
        Log.d("PROVA", "executing query");
        databaseReference.child(CHILD_MATCHID).child(CHILD_PLAYERS).addListenerForSingleValueEvent(new ValueEventListener() {

            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                for (DataSnapshot singleDataSnapshot : dataSnapshot.getChildren()) {
                    HashMap<String, String> player_info = (HashMap<String, String>) singleDataSnapshot.getValue();
                    Log.d("PROVA", player_info.toString());
                    playerID = singleDataSnapshot.getKey();
                    Log.d("PLAYERID", playerID);

                    if (!Boolean.parseBoolean(player_info.get("taken"))) {
                        playerOwnColor = player_info.get("color");

                        render();
                        databaseReference.child(CHILD_MATCHID).child(CHILD_PLAYERS).child(playerID).child("taken").setValue("true");
                        buttonColor.setText(playerID + " " + playerOwnColor);
                        Log.d("CHILDSNAPSHOT", "changing value");
                        break;
                    }


                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });
    }

    private void render() {
        if (playerOwnColor.equals("YELLOW")) {
            buttonColor.getBackground().setColorFilter(ContextCompat.getColor(this, R.color.myYellow), PorterDuff.Mode.MULTIPLY);
        }
        if (playerOwnColor.equals("GREEN")) {
            buttonColor.getBackground().setColorFilter(ContextCompat.getColor(this, R.color.myGreen), PorterDuff.Mode.MULTIPLY);
        }
        if (playerOwnColor.equals("RED")) {
            buttonColor.getBackground().setColorFilter(ContextCompat.getColor(this, R.color.myRed), PorterDuff.Mode.MULTIPLY);

        }
        if (playerOwnColor.equals("BLUE")) {
            buttonColor.getBackground().setColorFilter(ContextCompat.getColor(this, R.color.myBlue), PorterDuff.Mode.MULTIPLY);

        }

    }

    private void blink() {
        final DatabaseReference cpuSequenceRef = databaseReference.child(CHILD_MATCHID).child(CHILD_CPUSEQUENCE).getRef();
        databaseReference.child(CHILD_MATCHID).child(CHILD_INDEX).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                Log.d("PROVAINDEX_BLINK", dataSnapshot.getValue(String.class));
                final String cpuSequenceIndex = dataSnapshot.getValue(String.class);
                cpuSequenceRef.addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        String childrenCount = String.valueOf(dataSnapshot.getChildrenCount());
                        for (DataSnapshot child : dataSnapshot.getChildren()) {
                            String colorSequence = child.getValue(String.class);
                            String index = child.getKey();
                            Log.d("CHILDLOOP", colorSequence + " " + index);

                            if (playerOwnColor.equals(colorSequence) && cpuSequenceIndex.equals(index)) {
                                ++blinkCount;
                                Log.d("BLINKING", String.valueOf(blinkCount));
                                if (index.equals(childrenCount)) {

                                    buttonColor.setText(playerOwnColor + " " + blinkCount + " your turn!");

                                } else {
                                    int newIndex = Integer.parseInt(index) + 1;
                                    buttonColor.setText(playerOwnColor + " " + blinkCount);
                                    databaseReference.child(CHILD_MATCHID).child(CHILD_INDEX).setValue(String.valueOf(newIndex));

                                }


                            } else {
                                buttonColor.setText(playerOwnColor + " " + blinkCount);

                            }

                        }

                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {

                    }
                });
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

    }
}

File: app/src/main/java/app/simone/multiplayer/view/nearby/ColorSetUpActivity.kt
package app.simone.multiplayer.view.nearby


import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.Button
import app.simone.R
import app.simone.shared.utils.AudioPlayer
import app.simone.singleplayer.model.SColor
import com.facebook.Profile
import com.google.firebase.database.*
import java.util.concurrent.ExecutionException

class ColorSetUpActivity : AppCompatActivity() {

    private var playerID = ""
    private var matchID = ""
    private var playerColor: SColor? = null
    private var db: DatabaseReference? = null
    private val CHILD_PLAYERS = "users"
    private val NODE_REF_ROOT = "matches"
    private val CHILD_PLAYERSSEQUENCE = "playersSequence"
    private val CHILD_CPUSEQUENCE = "cpuSequence"
    private val CHILD_INDEX = "index"
    private var cpuSequenceRef : DatabaseReference? = null
    private var buttonColor: Button? = null
    private var blinkCount = 0
    private var player = AudioPlayer()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_color_set_up)
        db = FirebaseDatabase.getInstance().getReference(NODE_REF_ROOT)
        buttonColor = findViewById(R.id.beautifulButton) as Button

        matchID = intent.getStringExtra("match")
        playerID = Profile.getCurrentProfile().id

        setColor()
        observePlayersSequence()
    }

    public override fun onResume() {
        super.onResume()
        blink()
    }

    @Throws(ExecutionException::class, InterruptedException::class)
    fun sendColor(view: View) {

        if(playerColor != null) {
            player.play(this, playerColor!!.soundId)
        }

        db?.child(matchID)?.child(CHILD_PLAYERSSEQUENCE)?.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if(dataSnapshot.value != "playing") {
                    val count = dataSnapshot.childrenCount
                    dataSnapshot.ref.child((count + 1).toString()).setValue(playerColor)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) { }
        })

    }

    private fun setColor() {
        Log.d("PROVA", "executing query")
        db?.child(matchID)?.child(CHILD_PLAYERS)?.child(playerID)?.addValueEventListener(object : ValueEventListener {

            override fun onDataChange(dataSnapshot: DataSnapshot) {

                val child = dataSnapshot.child("color")

                if(child.value != null) {
                    val color = child.value.toString()
                    playerColor = SColor.valueOf(color)
                    render()
                }
            }

            override fun onCancelled(databaseError: DatabaseError) { }
        })
    }

    private fun observePlayersSequence() {

        /*
        db?.child(matchID)?.child(CHILD_PLAYERSSEQUENCE)?.addValueEventListener(object: ValueEventListener {
            override fun onCancelled(p0: DatabaseError?) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onDataChange(p0: DataSnapshot?) {
                val value = p0?.value

                if(value == null || value == "wrong") {

                }

            }

        })*/

        db?.child(matchID)?.child("blink")?.addValueEventListener(object: ValueEventListener{
            override fun onCancelled(p0: DatabaseError?) {

            }

            override fun onDataChange(p0: DataSnapshot?) {
                if(p0?.child("color")?.value == playerColor.toString()) {
                    renderBlink(p0.child("index")?.value.toString())
                }
            }

        })
    }

    private fun renderBlink(sequenceIndex: String) {
        buttonColor?.alpha = 0.5F
        player.play(this, playerColor!!.soundId)

        val handler = Handler()
        handler.postDelayed(Runnable {
            buttonColor?.alpha = 1.0F

            //db?.child(matchID)?.child("blink")?.child("status")?.setValue("replied")

            val nextIndex = sequenceIndex.toLong() + 1

            val cpuSeqRef = db?.child(matchID)?.child(CHILD_CPUSEQUENCE)?.ref

            cpuSeqRef?.addListenerForSingleValueEvent(object : ValueEventListener{
                override fun onCancelled(p0: DatabaseError?) {
                }

                override fun onDataChange(p0: DataSnapshot?) {
                    if(p0 != null && nextIndex <= p0.childrenCount) {
                        val nextColor = p0?.child(nextIndex.toString()).value.toString()

                        val map = HashMap<String,String>()
                        map["color"] = nextColor
                        map["index"] = nextIndex.toString()
                        db?.child(matchID)?.child("blink")?.setValue(map)

                        if(nextIndex == p0.childrenCount) {
                            db?.child(matchID)?.child(CHILD_PLAYERSSEQUENCE)?.setValue("empty")
                        }
                    }
                }
            })

        }, 500)
    }

    private fun render() {
        if(playerColor != null) {
            buttonColor?.background = resources?.getDrawable(playerColor!!.colorId)
        }
    }

    private fun blink() {
        cpuSequenceRef = db?.child(matchID)?.child(CHILD_CPUSEQUENCE)?.ref
        db?.child(matchID)?.child(CHILD_INDEX)?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                //Log.d("PROVAINDEX_BLINK", dataSnapshot.getValue(String.class));
                val index = dataSnapshot.getValue(String::class.java)
                if(index != null) {
                    observeCpuSequence(index)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) { }
        })

    }

    private fun observeCpuSequence(cpuIndex: String) {
        cpuSequenceRef?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val childrenCount = dataSnapshot.childrenCount.toString()
                for (child in dataSnapshot.children) {
                    val colorSequence = child.getValue(String::class.java)
                    val index = child.key
                    //Log.d("CHILDLOOP", colorSequence + " " + index);

                    if (colorSequence != null
                            && playerColor == SColor.valueOf(colorSequence)
                            && cpuIndex == index) {
                        ++blinkCount

                        if (index == childrenCount) {
                            buttonColor?.text = playerColor.toString() + " " + blinkCount + " your turn!"
                        } else {
                            val newIndex = Integer.parseInt(index) + 1
                            buttonColor?.text = playerColor.toString() + " " + blinkCount
                            db?.child(matchID)?.child(CHILD_INDEX)?.setValue(newIndex.toString())
                        }
                    } else {
                        buttonColor?.text = playerColor.toString() + " " + blinkCount
                    }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) { }
        })
    }

    companion object {
        private val handler: android.os.Handler? = null
    }
}
